clients/da-vinci-client/src/main/java/com/linkedin/davinci/blob/BlobSnapshotManager.java

```diff
@@ -0,0 +1,116 @@
+    if (fullPathForPartitionDBSnapshot.isEmpty()) {
+      return;
+    }
+    try {
+      Checkpoint checkpoint = createCheckpoint(rocksDB);
+
+      LOGGER.info("Start creating snapshots in directory: {}", fullPathForPartitionDBSnapshot);
+      checkpoint.createCheckpoint(fullPathForPartitionDBSnapshot);
+
+      LOGGER.info("Finished creating snapshots in directory: {}", fullPathForPartitionDBSnapshot);
+    } catch (RocksDBException e) {
+      throw new VeniceException(
+          "Received exception during RocksDB's snapshot creation in directory " + fullPathForPartitionDBSnapshot,
+          e);
+    }
+  }
+
+  // Checks if the snapshot is stale, if it is and no one is using it, it updates the snapshot, otherwise it increases
+  // the count of people using the snapshot
+  public void getHybridSnapshot(RocksDB rocksDB, String topicName, int partitionId) {
+    if (rocksDB == null || topicName == null) {
+      throw new IllegalArgumentException("RocksDB instance and topicName cannot be null");
+    }
+
+    String fullPathForPartitionDBSnapshot = RocksDBUtils.composeSnapshotDir(this.basePath, topicName, partitionId);
+    if (!snapShotTimestamps.containsKey(topicName)) {
+      snapShotTimestamps.put(topicName, new HashMap<>());
+      concurrentSnapshotUsers.put(topicName, new HashMap<>());
+    }
+    if (isSnapshotStale(topicName, partitionId)) {
+      if (concurrentSnapshotUsers.get(topicName).get(partitionId) == null
+          || concurrentSnapshotUsers.get(topicName).get(partitionId).get() == 0) {
+        updateHybridSnapshot(rocksDB, topicName, partitionId);
+        snapShotTimestamps.get(topicName).put(partitionId, System.currentTimeMillis());
+      }
+      if (getConcurrentUsers(rocksDB, topicName, partitionId) == 0) {
+        concurrentSnapshotUsers.get(topicName).put(partitionId, new AtomicLong(1));
+      } else {
+        concurrentSnapshotUsers.get(topicName).get(partitionId).incrementAndGet();
+      }
+      return;
+    }
+    try {
+      if (concurrentSnapshotUsers.get(topicName).get(partitionId) != null
+          && concurrentSnapshotUsers.get(topicName).get(partitionId).get() >= 0) {
+        concurrentSnapshotUsers.get(topicName).get(partitionId).incrementAndGet();
+      } else {
+        concurrentSnapshotUsers.get(topicName).put(partitionId, new AtomicLong(1));
+      }
+      LOGGER.info("Retrieved preexisting snapshot in directory: {}", fullPathForPartitionDBSnapshot);
+    } catch (Exception e) {
+      throw new VeniceException("Error getting snapshot", e);
+    }
+  }
+
+  public int getConcurrentUsers(RocksDB rocksDB, String topicName, int partitionId) {
+    if (rocksDB == null || topicName == null) {
+      throw new IllegalArgumentException("RocksDB instance and topicName cannot be null");
+    }
+    if (!concurrentSnapshotUsers.containsKey(topicName)) {
+      return 0;
+    }
+    if (!concurrentSnapshotUsers.get(topicName).containsKey(partitionId)) {
+      return 0;
+    }
```

Comment:
nit: These two cases could be combined. But also, it would probably be better to create a helper for this?

Reply:
Combined them since its only used in this case